### ISSUE TITLE
Issues about LINE

### ISSUE THREAD

**Issue by AndyJZhao** [iamlockelightning/NetworkEmbeddingPytorch#1]
Hi Locke, thanks for the open-source code, I really appreciate that. However, the loss of LINE seems not to descend in my run.

If I run LINE.py, during the training process of the LINE model with default parameters (batch_num=50000, batch_size=100, dataset='cora', gpu=True, negative=5, order=2, output='workspace/line_embedding_cora', path='data/cora/', rho=0.025, size=128),
the loss remain unchanged. (loss =0.6931) After the LINE embedding is generated, I run LRclassifier.py with result (Test set results: loss= 1.8627 accuracy= 0.3090).

I've only changed the code by replacing ".data[0]" to ".item()" for Pytorch version issues, other parts of the code remain unchanged.

Looking forward to your reply! Thanks a lot.

**Comment by iamlockelightning** [comment]
Hi Andy, thank you for using this code. 

This problem may be caught by the optimizer. You can try a different optimizer instead of using Adadelta.

For example, I changed the LINE.py to Pytorch 0.4 version and changed the optimizer to Adagrad, I got the following loss change, which seemed normal. But you can also try different parameters for good results. I am sorry that I don't have much time to pursue good results with you further.

```
Namespace(batch_num=1000, batch_size=100, dataset='cora', negative=5, order=2, output='workspace/line_embedding_cora', path='data/cora/', rho=0.025, size=128)
Loading cora dataset...
Dataset has 2708 nodes, 5429 edges, 1433 features.
Batch_no: 000100 loss: 0.6800 rho: 0.0250 sampling_time: 3.2896 training_time: 0.1041
Batch_no: 000200 loss: 0.6671 rho: 0.0250 sampling_time: 2.9714 training_time: 0.0802
Batch_no: 000300 loss: 0.6562 rho: 0.0250 sampling_time: 2.7844 training_time: 0.0902
Batch_no: 000400 loss: 0.6444 rho: 0.0250 sampling_time: 2.9740 training_time: 0.0985
Batch_no: 000500 loss: 0.6455 rho: 0.0250 sampling_time: 2.9056 training_time: 0.0971
Batch_no: 000600 loss: 0.6406 rho: 0.0250 sampling_time: 3.0760 training_time: 0.0858
Batch_no: 000700 loss: 0.6327 rho: 0.0250 sampling_time: 2.7918 training_time: 0.0846
Batch_no: 000800 loss: 0.6337 rho: 0.0250 sampling_time: 2.9611 training_time: 0.0743
Batch_no: 000900 loss: 0.6356 rho: 0.0250 sampling_time: 3.2736 training_time: 0.0845
done..
```

The changed LINE.py is in the attachment.
[LINE_pytorch0.4.py.zip](https://github.com/iamlockelightning/NetworkEmbeddingPytorch/files/3528157/LINE_pytorch0.4.py.zip)

Pull request are welcomed to make all code available in higher versions of Pytorch!

**Comment by iamlockelightning** [comment]
Hi @Andy-Border , thank you for using this code. 

This problem may be caused by the optimizer. You can try a different optimizer instead of using Adadelta.

For example, I changed the LINE.py to Pytorch 0.4 version and changed the optimizer to Adagrad, I got the following loss change, which seemed normal. But you can also try different parameters for good results. I am sorry that I don't have much time to pursue good results with you further.

```
Namespace(batch_num=1000, batch_size=100, dataset='cora', negative=5, order=2, output='workspace/line_embedding_cora', path='data/cora/', rho=0.025, size=128)
Loading cora dataset...
Dataset has 2708 nodes, 5429 edges, 1433 features.
Batch_no: 000100 loss: 0.6800 rho: 0.0250 sampling_time: 3.2896 training_time: 0.1041
Batch_no: 000200 loss: 0.6671 rho: 0.0250 sampling_time: 2.9714 training_time: 0.0802
Batch_no: 000300 loss: 0.6562 rho: 0.0250 sampling_time: 2.7844 training_time: 0.0902
Batch_no: 000400 loss: 0.6444 rho: 0.0250 sampling_time: 2.9740 training_time: 0.0985
Batch_no: 000500 loss: 0.6455 rho: 0.0250 sampling_time: 2.9056 training_time: 0.0971
Batch_no: 000600 loss: 0.6406 rho: 0.0250 sampling_time: 3.0760 training_time: 0.0858
Batch_no: 000700 loss: 0.6327 rho: 0.0250 sampling_time: 2.7918 training_time: 0.0846
Batch_no: 000800 loss: 0.6337 rho: 0.0250 sampling_time: 2.9611 training_time: 0.0743
Batch_no: 000900 loss: 0.6356 rho: 0.0250 sampling_time: 3.2736 training_time: 0.0845
done..
```

The changed LINE.py is in the attachment.
[LINE_pytorch0.4.py.zip](https://github.com/iamlockelightning/NetworkEmbeddingPytorch/files/3528157/LINE_pytorch0.4.py.zip)

Pull requests are welcomed to make all code available in higher versions of Pytorch!
